charms/worker/k8s/src/charm.py

```diff
@@ -383,39 +383,62 @@ def _get_valid_annotations(self) -> Optional[dict]:
-        """Configure the annotations for the Canonical Kubernetes cluster.
+    def _get_cluster_config_from_charm_config(self) -> UserFacingClusterConfig:
+        """Retrieve the cluster config from charm configuration.
+
+        Returns:
+            UserFacingClusterConfig: The expected cluster configuration.
+        """
+        local_storage = LocalStorageConfig(
+            enabled=self.config.get("local-storage-enabled"),
+            local_path=self.config.get("local-storage-local-path"),
+            reclaim_policy=self.config.get("local-storage-reclaim-policy"),
+            # Note(ben): set_default is intentionally omitted, see:
+            # https://github.com/canonical/k8s-operator/pull/169/files#r1847378214
+        )
+
+        return UserFacingClusterConfig(
+            local_storage=local_storage,
+            annotations=self._get_valid_annotations(),
+        )
+
+    def _configure_cluster_config(self, config: BootstrapConfig):
+        """Configure the cluster config for the Canonical Kubernetes cluster.
 
         Args:
             config (BootstrapConfig): Configuration object to bootstrap the cluster.
         """
-        annotations = self._get_valid_annotations()
-        if annotations is None:
-            return
-
         if not config.cluster_config:
-            config.cluster_config = UserFacingClusterConfig(annotations=annotations)
+            config.cluster_config = self._get_cluster_config_from_charm_config()
         else:
-            config.cluster_config.annotations = annotations
+            charm_cluster_config = self._get_cluster_config_from_charm_config()
+            # Merge the charm and user cluster configurations
+            # The existing configuration takes precedence over the charm configuration as
+            # this information is retrieved elsewhere (e.g. cloud-provider).
+            merged_data = {**charm_cluster_config.dict(), **config.cluster_config.dict()}
+            config.cluster_config = UserFacingClusterConfig(**merged_data)
 
     @status.on_error(
-        ops.BlockedStatus("Invalid Annotations"),
+        ops.BlockedStatus("Invalid Cluster configuration"),
         AssertionError,
     )
-    def _update_annotations(self):
-        """Update the annotations for the Canonical Kubernetes cluster."""
-        annotations = self._get_valid_annotations()
-        if annotations is None:
-            return
+    def _update_cluster_config(self):
+        """Update the cluster configuration for the Canonical Kubernetes cluster."""
+        status.add(ops.MaintenanceStatus("Updating cluster configuration"))
+        log.info("Updating cluster configuration")
 
-        status.add(ops.MaintenanceStatus("Updating Annotations"))
-        log.info("Updating Annotations")
+        charm_cluster_config = self._get_cluster_config_from_charm_config()
+        snap_cluster_config = self.api_manager.get_cluster_status().metadata.status.config
```

Comment:
Seems like `.metadata` can be `None`. Maybe we can check that to prevent exceptions?

Reply:
I think that should be fine and is as designed.
The reconciler loop will just fail and retry, am I correct @mateoflorido ?

Reply:
That being said, it is outdated now anyway - we can safely send the update request to the snap and the snap will only perform any actions if something actually changed.